File: src/cubyz/world/terrain/generators/OreGenerator.java
package cubyz.world.terrain.generators;

import cubyz.api.CurrentWorldRegistries;
import cubyz.api.Resource;
import cubyz.utils.FastRandom;
import cubyz.world.Chunk;
import cubyz.world.NormalChunk;
import cubyz.world.blocks.Blocks;
import cubyz.world.blocks.Ore;
import cubyz.world.terrain.CaveBiomeMap;
import cubyz.world.terrain.CaveMap;
import pixelguys.json.JsonObject;

/**
 * Generator of ore veins.
 */

public class OreGenerator implements Generator {
	
	@Override
	public Resource getRegistryID() {
		return new Resource("cubyz", "ore");
	}
	
	@Override
	public int getPriority() {
		return 32768; // Somewhere before cave generation.
	}

	private Ore[] ores;

	@Override
	public void init(JsonObject parameters, CurrentWorldRegistries registries) {
		ores = registries.oreRegistry.registered(new Ore[0]);
	}


	// Works basically similar to cave generation, but considers a lot less chunks and has a few other differences.
	@Override
	public void generate(long seed, int wx, int wy, int wz, Chunk chunk, CaveMap caveMap, CaveBiomeMap biomeMap) {
		if (!(chunk instanceof NormalChunk)) return;
		FastRandom rand = new FastRandom(seed);
		int rand1 = rand.nextInt() | 1;
		int rand2 = rand.nextInt() | 1;
		int rand3 = rand.nextInt() | 1;
		int cx = wx >> NormalChunk.chunkShift;
		int cy = wy >> NormalChunk.chunkShift;
		int cz = wz >> NormalChunk.chunkShift;
		// Generate caves from all nearby chunks:
		for(int x = cx - 1; x <= cx + 1; ++x) {
			for(int y = cy - 1; y <= cy + 1; ++y) {
				for(int z = cz - 1; z <= cz + 1; ++z) {
					int randX = x*rand1;
					int randY = y*rand2;
					int randZ = z*rand3;
					considerCoordinates(x, y, z, cx, cy, cz, chunk, (randY << 48) ^ (randY >>> 16) ^ (randX << 32) ^ randZ ^ seed);
				}
			}
		}
	}

	private void considerCoordinates(int x, int y, int z, int cx, int cy, int cz, Chunk chunk, long seed) {
		FastRandom rand = new FastRandom(0);
		for(Ore ore : ores) {
			if(ore.maxHeight <= y << NormalChunk.chunkShift) continue;
			// Compose the seeds from some random stats of the ore. They generally shouldn't be the same for two different ores.
			rand.setSeed(seed ^ (ore.maxHeight) ^ (Float.floatToIntBits(ore.size)) ^ Blocks.id(ore.block).getID().charAt(0) ^ Float.floatToIntBits(Blocks.hardness(ore.block)));
			// Determine how many veins of this type start in this chunk. The number depends on parameters set for the specific ore:
			int veins = (int)ore.veins;
			if(ore.veins - veins >= rand.nextFloat()) veins++;
			for(int j = 0; j < veins; ++j) {
				// Choose some in world coordinates to start generating:
				double relX = (x-cx << NormalChunk.chunkShift) + rand.nextFloat()*NormalChunk.chunkSize;
				double relY = (y-cy << NormalChunk.chunkShift) + rand.nextFloat()*NormalChunk.chunkSize;
				double relZ = (z-cz << NormalChunk.chunkShift) + rand.nextFloat()*NormalChunk.chunkSize;
				// Choose a random volume and create a radius from that:
				double size = (rand.nextFloat() + 0.5f)*ore.size;
				double expectedVolume = 2*size/ore.density; // Double the volume, because later the density is actually halfed.
				double radius = Math.cbrt(expectedVolume*3/4/Math.PI);
				int xMin = (int)Math.ceil(relX - radius);
				int xMax = (int)Math.ceil(relX + radius);
				int zMin = (int)Math.ceil(relZ - radius);
				int zMax = (int)Math.ceil(relZ + radius);
				xMin = Math.max(xMin, 0);
				xMax = Math.min(xMax, Chunk.chunkSize*chunk.voxelSize);
				zMin = Math.max(zMin, 0);
				zMax = Math.min(zMax, Chunk.chunkSize*chunk.voxelSize);
				FastRandom noiseRand = new FastRandom(rand.nextLong());

				for(int curX = xMin; curX < xMax; curX += chunk.voxelSize) {
					double distToCenterX = (curX - relX)/radius;
					for(int curZ = zMin; curZ < zMax; curZ += chunk.voxelSize) {
						double distToCenterZ = (curZ - relZ)/radius;
						double yDistance = radius*Math.sqrt(1 - distToCenterX*distToCenterX - distToCenterZ*distToCenterZ);
						int yMin = (int)Math.ceil(relY - yDistance);
						int yMax = (int)Math.ceil(relY + yDistance);
						yMin = Math.max(yMin, 0);
						yMax = Math.min(yMax, Chunk.chunkSize*chunk.voxelSize);

						for(int curY = yMin; curY < yMax; curY += chunk.voxelSize) {
							double distToCenterY = (curY - relY)/radius;
							double distToCenter = distToCenterX*distToCenterX + distToCenterY*distToCenterY + distToCenterZ*distToCenterZ;
							if(distToCenter < 1) {
								// Add some roughness. The ore density gets smaller at the edges:
								if((1 - (distToCenter))*ore.density >= noiseRand.nextFloat()) {
									if(ore.canCreateVeinInBlock(chunk.getBlock(curX, curY, curZ))) {
										chunk.updateBlockInGeneration(curX, curY, curZ, ore.block);
									}
								}
							}
						}
					}
				}
			}
		}
	}

	@Override
	public long getGeneratorSeed() {
		return 0x88773787bc9e0105L;
	}
}

File: src/server/terrain/chunkgen/OreGenerator.zig
const std = @import("std");

const main = @import("main");
const random = main.random;
const ZonElement = main.ZonElement;
const terrain = main.server.terrain;
const CaveMap = terrain.CaveMap;
const CaveBiomeMap = terrain.CaveBiomeMap;
const vec = main.vec;
const Vec3d = vec.Vec3d;
const Vec3f = vec.Vec3f;
const Vec3i = vec.Vec3i;

pub const id = "cubyz:ore";

pub const priority = 32768;

pub const generatorSeed = 0x88773787bc9e0105;

var ores: []main.blocks.Ore = undefined;

// TODO: Idea:
// Add a RotationMode that allows you to overlay the ore texture onto a regular block to get more ore-in-stone-types for free.

pub fn init(parameters: ZonElement) void {
	_ = parameters;
	ores = main.blocks.ores.items;
}

// Works basically similar to cave generation, but considers a lot less chunks and has a few other differences.
pub fn generate(worldSeed: u64, chunk: *main.chunk.ServerChunk, caveMap: CaveMap.CaveMapView, biomeMap: CaveBiomeMap.CaveBiomeMapView) void {
	_ = caveMap;
	_ = biomeMap;
	if(chunk.super.pos.voxelSize != 1) return;
	const cx = chunk.super.pos.wx >> main.chunk.chunkShift;
	const cy = chunk.super.pos.wy >> main.chunk.chunkShift;
	const cz = chunk.super.pos.wz >> main.chunk.chunkShift;
	// Generate caves from all nearby chunks:
	var x = cx - 1;
	while(x < cx + 1) : (x +%= 1) {
		var y = cy - 1;
		while(y < cy + 1) : (y +%= 1) {
			var z = cz - 1;
			while(z < cz + 1) : (z +%= 1) {
				const seed = random.initSeed3D(worldSeed, .{x, y, z});
				const relX: f32 = @floatFromInt(x - cx << main.chunk.chunkShift);
				const relY: f32 = @floatFromInt(y - cy << main.chunk.chunkShift);
				const relZ: f32 = @floatFromInt(z - cz << main.chunk.chunkShift);
				for(ores) |*ore| {
					if(ore.maxHeight <= z << main.chunk.chunkShift or ore.minHeight > z << main.chunk.chunkShift) continue;
					considerCoordinates(ore, relX, relY, relZ, chunk, seed);
				}
			}
		}
	}
}

fn considerCoordinates(ore: *const main.blocks.Ore, relX: f32, relY: f32, relZ: f32, chunk: *main.chunk.ServerChunk, startSeed: u64) void {
	const chunkSizeFloat: f32 = @floatFromInt(main.chunk.chunkSize);
	// Compose the seeds from some random stats of the ore. They generally shouldn't be the same for two different ores. TODO: Give each block a hash function (id based) that can be used in cases like this.
	var seed = startSeed ^ @as(u32, @bitCast(ore.maxHeight)) ^ @as(u32, @bitCast(ore.size)) ^ @as(u32, @bitCast(main.blocks.Block.blockHealth(.{.typ = ore.blockType, .data = 0})));
	random.scrambleSeed(&seed);
	// Determine how many veins of this type start in this chunk. The number depends on parameters set for the specific ore:
	const veins: u32 = @intFromFloat(random.nextFloat(&seed)*ore.veins*2);
	for(0..veins) |_| {
		// Choose some in world coordinates to start generating:
		const veinRelX = relX + random.nextFloat(&seed)*chunkSizeFloat;
		const veinRelY = relY + random.nextFloat(&seed)*chunkSizeFloat;
		const veinRelZ = relZ + random.nextFloat(&seed)*chunkSizeFloat;
		// Choose a random volume and create a radius from that:
		const size = (random.nextFloat(&seed) + 0.5)*ore.size;
		const expectedVolume = 2*size/ore.density; // Double the volume, because later the density is actually halfed.
		const radius = std.math.cbrt(expectedVolume*3/4/std.math.pi);
		var xMin: i32 = @intFromFloat(veinRelX - radius);
		var xMax: i32 = @intFromFloat(@ceil(veinRelX + radius));
		var yMin: i32 = @intFromFloat(veinRelY - radius);
		var yMax: i32 = @intFromFloat(@ceil(veinRelY + radius));
		xMin = @max(xMin, 0);
		xMax = @min(xMax, chunk.super.width);
		yMin = @max(yMin, 0);
		yMax = @min(yMax, chunk.super.width);

		var veinSeed = random.nextInt(u64, &seed);
		var curX = xMin;
		while(curX < xMax) : (curX += 1) {
			const distToCenterX = (@as(f32, @floatFromInt(curX)) - veinRelX)/radius;
			var curY = yMin;
			while(curY < yMax) : (curY += 1) {
				const distToCenterY = (@as(f32, @floatFromInt(curY)) - veinRelY)/radius;
				const xyDistSqr = distToCenterX*distToCenterX + distToCenterY*distToCenterY;
				if(xyDistSqr > 1) continue;
				const zDistance = radius*@sqrt(1 - xyDistSqr);
				var zMin: i32 = @intFromFloat(veinRelZ - zDistance);
				var zMax: i32 = @intFromFloat(@ceil(veinRelZ + zDistance));
				zMin = @max(zMin, 0);
				zMax = @min(zMax, chunk.super.width);
				var curZ = zMin;
				while(curZ < zMax) : (curZ += 1) {
					const distToCenterZ = (@as(f32, @floatFromInt(curZ)) - veinRelZ)/radius;
					const distSqr = xyDistSqr + distToCenterZ*distToCenterZ;
					if(distSqr < 1) {
						// Add some roughness. The ore density gets smaller at the edges:
						if((1 - distSqr)*ore.density >= random.nextFloat(&veinSeed)) {
							const stoneBlock = chunk.getBlock(curX, curY, curZ);
							if(chunk.getBlock(curX, curY, curZ).allowOres()) {
								chunk.updateBlockInGeneration(curX, curY, curZ, .{.typ = ore.blockType, .data = stoneBlock.typ});
							}
						}
					}
				}
			}
		}
	}
}
